### PR TITLE
fix: refresh buffered chunk tokens before OM activation

### PR DESCRIPTION
**Pull request by TylerBarnes**
Fixes observational memory activation selecting too many buffered chunks by refreshing chunk token counts from the current message list before activation math runs. Previously, stale token weights from buffering time could cause over- or under-activation.

Also adds cursor-based replay pruning so already-observed messages are filtered on all steps (not just step 0), with a `didThresholdCleanup` guard to avoid double-pruning when threshold cleanup already ran.

**Core fix**: `refreshBufferedChunkMessageTokens()` recomputes per-chunk `messageTokens` from live context before projected-removal math. Only recounts when the full chunk is present — partial recount is skipped to avoid undercounting.

**Testing**: 275/275 OM tests passing, including cherry-picked regression tests from the original branch with adjusted thresholds for the recounting fix.

Changes span `@mastra/memory` (main fix) and `@mastra/core` (storage plumbing for `bufferedChunks` override + `lastObservedMessageCursor` type).

## Summary by CodeRabbit

* **Bug Fixes**
  * Observational memory activation now uses up-to-date thread state and cursor-based pruning to prevent stale token counts and improve recall accuracy.

* **Tests**
  * Added extensive single-thread replay tests covering boundary pruning, activation timing, replay behavior, and related edge cases.

* **Chores**
  * Added a changelog entry documenting the patch release fix.

### CODE REVIEW

**Review by abhiaiyer91 — APPROVED** [review]

**Review by DanielSLew — APPROVED** [review]

File: packages/core/src/memory/types.ts [hunk 1]
```diff
@@ -57,6 +57,8 @@ export type ThreadOMMetadata = {
   suggestedResponse?: string;
   /** Timestamp of the last observed message in this thread (ISO string for JSON serialization) */
   lastObservedAt?: string;
+  /** Cursor pointing at the last observed message (for replay pruning fallback) */
+  lastObservedMessageCursor?: { createdAt: string; id: string };
```
[the remark this answers is not in the document]
Comment (TylerBarnes) [reply packages/core/src/memory/types.ts:61]: Reese says: `getThreadOMMetadata` is an internal helper, not a system boundary — this metadata is written and read exclusively within our own OM processor. Adding runtime shape validation here would be defensive coding against ourselves with no real benefit. If the cursor shape is wrong, we have bigger problems upstream. Leaving this as-is.

Cheers, Claude

[the remark this answers is not in the document]
Comment (TylerBarnes) [reply packages/core/src/memory/types.ts:61]: Reese says: `getThreadOMMetadata` is an internal helper, not a system boundary — this metadata is written and read exclusively within our own OM processor. Adding runtime shape validation here would be defensive coding against ourselves with no real benefit. If the cursor shape is wrong, we have bigger problems upstream. Leaving this as-is.

Cheers, Claude

File: packages/memory/src/processors/observational-memory/observational-memory.ts [hunk 2]
```diff
@@ -3134,10 +3233,14 @@ ${suggestedResponse}
...
+    // - step 0: use marker-boundary pruning + record fallback (historical resume)
+    // - step >0: use record fallback only (avoid position-based marker over-pruning mid-loop)
     // ════════════════════════════════════════════════════════════════════════
-    if (stepNumber === 0) {
-      this.filterAlreadyObservedMessages(messageList, record);
+    // If step-level cleanup already ran after threshold handling, skip this pass to avoid
+    // a second timestamp-based prune that can undercut retention-floor guarantees.
+    if (!didThresholdCleanup) {
+      await this.filterAlreadyObservedMessages(messageList, record, { useMarkerBoundaryPruning: stepNumber === 0 });
     }
```
[the remark this answers is not in the document]
Comment (TylerBarnes) [reply packages/memory/src/processors/observational-memory/observational-memory.ts:3244]: Reese says: This is safe by design. The concern is that `observedMessageIds` in the record-fallback branch could remove tail content preserved by `cleanupAfterObservation`. But `swapBufferedToActive` intentionally does NOT add activated message IDs to `record.observedMessageIds` — activated messages are removed from context immediately via `messageList.removeByIds()` during cleanup, not via the observation record. So the fallback branch in `filterAlreadyObservedMessages` will never see those IDs in `observedMessageIds` and cannot re-remove them.

Additionally, `didThresholdCleanup` guards the same-step case: threshold cleanup sets it true, preventing the filter from running again in the same `processInputStep` call. On subsequent steps, the messages are already gone from context (removed during activation cleanup), so there is nothing left to incorrectly prune.

Yours truly, Claude
